go/store/types/edits/tuple_stream_merger.go

```diff
@@ -0,0 +1,231 @@
+	readers := make([]types.TupleReadCloser, len(filenames))
+	for i, name := range filenames {
+		f, err := os.Open(name)
+		if err != nil {
+			return nil, err
+		}
+
+		readers[i] = types.NewTupleReader(nbf, vrw, f)
+	}
+
+	deleteFile := func(i int) {
+		if deleteOnExit {
+			path := filenames[i]
+			_ = os.Remove(path)
+		}
+	}
+
+	return NewTupleStreamMerger(ctx, nbf, readers, numEdits, deleteFile)
+}
+
+// NewTupleStreamMerger takes a slice of TupleReaders, whose contents should be key sorted key value tuple
+// pairs, and return a *TupleStreamMerger
+func NewTupleStreamMerger(ctx context.Context, nbf *types.NomsBinFormat, readers []types.TupleReadCloser, numEdits int64, readerClosed func(i int)) (*TupleStreamMerger, error) {
+	fep := &TupleStreamMerger{
+		ctx:          ctx,
+		nbf:          nbf,
+		numEdits:     numEdits,
+		numReaders:   len(readers),
+		readers:      readers,
+		nextKVPS:     make([]entry, 0, len(readers)),
+		readerClosed: readerClosed,
+	}
+
+	// read in the initial values from each stream and put them into the nextKVPS slice in sorted order.
+	for i := range readers {
+		kvp, err := fep.readKVP(i)
+		if err == io.EOF {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		// store the kvp along with the index of the reader it was read from.
+		newEntry := entry{kvp: kvp, readerIdx: i}
+
+		// binary search for where this entry should be inserted within the slice
+		insIdx, err := search(nbf, kvp.Key, fep.nextKVPS)
+		if err != nil {
+			return nil, err
+		}
+
+		// grow the slice of entries
+		fep.nextKVPS = fep.nextKVPS[:len(fep.nextKVPS)+1]
+
+		// if necessary move existing entries to make room for new entry to be inserted in the correct place
+		if insIdx < len(fep.nextKVPS)-1 {
+			copy(fep.nextKVPS[insIdx+1:], fep.nextKVPS[insIdx:len(fep.nextKVPS)-1])
+		}
+
+		fep.nextKVPS[insIdx] = newEntry
+	}
+
+	fep.readersWithData = len(fep.nextKVPS)
+
+	return fep, nil
+}
+
+// Next returns the next KVP representing the next edit to be applied.  Next will always return KVPs
+// in key sorted order.  Once all KVPs have been read io.EOF will be returned.
+func (fep *TupleStreamMerger) Next() (*types.KVP, error) {
+	if fep.readersWithData == 0 {
+		return nil, io.EOF
+	}
+
+	// nextKVP taken from index 0 as fep.nextKVPS are sorted low to high so element 0 is the next item
+	nextKVP := fep.nextKVPS[0]
+
+	// read the next tuple from the TupleStream that next kvp was read from
+	kvp, err := fep.readKVP(nextKVP.readerIdx)
+	if err == io.EOF {
+		// shrink the slice to only hold valid ordered data
+		fep.nextKVPS = fep.nextKVPS[1:]
+		fep.readersWithData--
+
+		// close the reader and execute close callback
+		fep.readers[nextKVP.readerIdx].Close(fep.ctx)
+
+		if fep.readerClosed != nil {
+			fep.readerClosed(nextKVP.readerIdx)
+		}
+	} else if err != nil {
+		return nil, err
+	} else {
+		// search for the location where the item should be placed
+		insPos, err := search(fep.nbf, kvp.Key, fep.nextKVPS[1:])
```

Comment:
suggestion: implement `sort.Interface` for `entry` and use `sort.Search` here

Reply:
Good idea I took your suggestion.  But sort.Search doesn't require you to implement an interface.

```go
func search(nbf *types.NomsBinFormat, key types.LesserValuable, vals []entry) (int, error) {
	var err error
	n := sort.Search(len(vals), func(i int) bool {
		if err != nil {
			return false
		}

		var isLess bool
		isLess, err = key.Less(nbf, vals[i].kvp.Key)
		return isLess
	})

	if err != nil {
		return 0, err
	}

	return n, nil
}
```